[Model Element] Delete model player when the HTMLModelElement has been stopped

achan00 opened the pull request:
#### 6caee26b090589e4939095c109751ca27b796075
<pre>
[Model Element] Delete model player when the HTMLModelElement has been stopped
<a href="https://bugs.webkit.org/show_bug.cgi?id=293817">https://bugs.webkit.org/show_bug.cgi?id=293817</a>
<a href="https://rdar.apple.com/151798302">rdar://151798302</a>

Reviewed by Mike Wyrzykowski.

Once the HTMLModelElement has been stopped (after a page reload for example),
it cannot be restarted. We can delete the model player at this point to free up
system resources.

Add plumbing to get the current count of ModelProcessModelPlayerProxy
for testing. Add API tests to make sure ModelProcessModelPlayerProxy is
released after page reloading, navigation, and becoming hidden.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::stop):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::modelPlayerCountForTesting):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _modelProcessModelPlayerCountForTesting:]):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::modelPlayerCountForTesting):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/simple-model-page.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ModelProcess.mm: Added.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnReload)):
After reloading a page with a single model, the number of model player
should stay at 1.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnNavigate)):
After navigating away from a page with a single model, there should be
no more model players.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnHide)):
After changing the page visibility to be hidden, there should be no more
model players.
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]): Deleted.
* Tools/TestWebKitAPI/cocoa/ModelLoadingMessageHandler.h: Added.
* Tools/TestWebKitAPI/cocoa/ModelLoadingMessageHandler.mm: Added.
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView modelProcessModelPlayerCount]):

Canonical link: <a href="https://commits.webkit.org/295706@main">https://commits.webkit.org/295706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6adb84b04c93730a1fc905eb05e555981439d557

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105724 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110921 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56319 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107765 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33978 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80288 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108730 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95406 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60598 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20108 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13500 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55759 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13539 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113770 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32864 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24223 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89367 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33228 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89032 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22739 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33908 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11712 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28324 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32789 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38199 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32535 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35884 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34133 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -207,6 +207,7 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
 
     // ActiveDOMObject.
     bool virtualHasPendingActivity() const final;
+    void stop() final;
```
Is it intentional that suspend() is public, but stop() is private?

achan00 replied:
Not intentional - they can all be private. Updated PR.

achan00 committed:
[Model Element] Delete model player when the HTMLModelElement has been stopped
https://bugs.webkit.org/show_bug.cgi?id=293817
rdar://151798302

Reviewed by Mike Wyrzykowski.

Once the HTMLModelElement has been stopped (after a page reload for example),
it cannot be restarted. We can delete the model player at this point to free up
system resources.

Add plumbing to get the current count of ModelProcessModelPlayerProxy
for testing. Add API tests to make sure ModelProcessModelPlayerProxy is
released after page reloading, navigation, and becoming hidden.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::stop):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebKit/ModelProcess/ModelProcess.cpp:
(WebKit::ModelProcess::modelPlayerCountForTesting):
* Source/WebKit/ModelProcess/ModelProcess.h:
* Source/WebKit/ModelProcess/ModelProcess.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _modelProcessModelPlayerCountForTesting:]):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::modelPlayerCountForTesting):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/simple-model-page.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ModelProcess.mm: Added.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnReload)):
After reloading a page with a single model, the number of model player
should stay at 1.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnNavigate)):
After navigating away from a page with a single model, there should be
no more model players.
(TestWebKitAPI::TEST(ModelProcess, CleanUpOnHide)):
After changing the page visibility to be hidden, there should be no more
model players.
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]): Deleted.
* Tools/TestWebKitAPI/cocoa/ModelLoadingMessageHandler.h: Added.
* Tools/TestWebKitAPI/cocoa/ModelLoadingMessageHandler.mm: Added.
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView modelProcessModelPlayerCount]):

Canonical link: https://commits.webkit.org/295706@main
